snap-geometry-edits/src/main/java/com/esri/arcgismaps/sample/snapgeometryedits/components/MapViewModel.kt

```diff
@@ -0,0 +1,249 @@
+
+import android.app.Application
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.ui.unit.dp
+import androidx.lifecycle.AndroidViewModel
+import com.arcgismaps.geometry.GeometryType
+import com.arcgismaps.geometry.Multipoint
+import com.arcgismaps.geometry.Point
+import com.arcgismaps.geometry.Polygon
+import com.arcgismaps.geometry.Polyline
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.layers.FeatureTilingMode
+import com.arcgismaps.mapping.view.Graphic
+import com.arcgismaps.mapping.view.GraphicsOverlay
+import com.arcgismaps.mapping.view.SingleTapConfirmedEvent
+import com.arcgismaps.mapping.view.geometryeditor.GeometryEditor
+import com.arcgismaps.mapping.view.geometryeditor.GeometryEditorStyle
+import com.arcgismaps.mapping.view.geometryeditor.ReticleVertexTool
+import com.arcgismaps.mapping.view.geometryeditor.SnapSourceSettings
+import com.arcgismaps.toolkit.geoviewcompose.MapViewProxy
+import com.esri.arcgismaps.sample.sampleslib.components.MessageDialogViewModel
+import com.esri.arcgismaps.sample.snapgeometryedits.R
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.launch
+
+class MapViewModel(
+    application: Application,
+    private val sampleCoroutineScope: CoroutineScope
+) : AndroidViewModel(application) {
+    // create a map using the URL of the web map
+    val map = ArcGISMap(application.getString(R.string.web_map))
+
+    // create a graphic, graphic overlay, and geometry editor
+    private var identifiedGraphic = Graphic()
+    val graphicsOverlay = GraphicsOverlay()
+    val geometryEditor = GeometryEditor()
+
+    // create a mapViewProxy that will be used to identify features in the MapView and set the viewpoint
+    val mapViewProxy = MapViewProxy()
+
+    // create a messageDialogViewModel to handle dialog interactions
+    val messageDialogVM: MessageDialogViewModel = MessageDialogViewModel()
+
+    // create lists for displaying the snap sources in the bottom sheet
+    private val _snapSourceSettingsList = MutableStateFlow(listOf<SnapSourceSettings>())
+    val snapSourceList: StateFlow<List<SnapSourceSettings>> = _snapSourceSettingsList
+
+    // create boolean flags to track the state of UI components
+    val isCreateButtonEnabled = mutableStateOf(false)
+    val isSnapSettingsButtonEnabled = mutableStateOf(false)
+    val isBottomSheetVisible = mutableStateOf(false)
+    val snappingCheckedState = mutableStateOf(false)
+    val snapSourceCheckedState = mutableStateListOf<Boolean>()
+    val isUndoButtonEnabled = geometryEditor.canUndo
+    val isSaveButtonEnabled = geometryEditor.isStarted
+    val isDeleteButtonEnabled = geometryEditor.isStarted
+
+    /**
+     * Configure the map and enable the UI after the map's layers are loaded.
+     */
+    init {
+        sampleCoroutineScope.launch {
```

Comment:
feels like this should be focused on just the load & success/failure block which is suspending